packages/remark-extend/src/remarkExtend.js

```diff
@@ -0,0 +1,291 @@
+    return true;
+  }
+  if (node.value && node.value.startsWith('::')) {
+    return true;
+  }
+  return false;
+}
+
+function findMdAdditionTasks(tree, file) {
+  let addNodes = [];
+  let gathering = false;
+  let startSelector;
+  let action = '';
+  visit(tree, (node, index, parent) => {
+    if (gathering === true && shouldFinishGathering(node)) {
+      gathering = false;
+      addTask(file, {
+        action,
+        startSelector,
+        addNodes,
+      });
+      addNodes = [];
+    }
+
+    if (gathering === true) {
+      if (parent.type === 'root') {
+        addNodes.push(node);
+      }
+    }
+
+    if (node.type === 'code' && node.value && node.value.startsWith('::addMdAfter')) {
+      startSelector = node.value.substring(node.value.indexOf("('") + 2, node.value.indexOf("')"));
+      gathering = true;
+      action = 'addMdAfter';
+    }
+    if (node.type === 'code' && node.value && node.value.startsWith('::addMdBefore')) {
+      startSelector = node.value.substring(node.value.indexOf("('") + 2, node.value.indexOf("')"));
+      gathering = true;
+      action = 'addMdBefore';
+    }
+  });
+
+  if (gathering === true) {
+    addTask(file, {
+      action,
+      startSelector,
+      addNodes,
+    });
+  }
+}
+
+function findRemoveTasks(tree, file) {
+  visit(tree, 'code', node => {
+    if (node.value && node.value.startsWith('::removeFrom')) {
+      const startSelector = node.value.substring(
+        node.value.indexOf("('") + 2,
+        node.value.indexOf("')"),
+      );
+      addTask(file, {
+        action: 'removeFrom',
+        startSelector,
+      });
+    }
+    if (node.value && node.value.startsWith('::removeBetween')) {
+      const startSelector = node.value.substring(
+        node.value.indexOf("('") + 2,
+        node.value.indexOf("',"),
+      );
+      const endSelector = node.value
+        .substring(node.value.indexOf("',") + 4, node.value.indexOf("')"))
+        .trim();
+      addTask(file, {
+        action: 'removeBetween',
+        startSelector,
+        endSelector,
+      });
+    }
+  });
+}
+
+function findExtendTasks() {
+  return (tree, file) => {
+    findReplacementTasks(tree, file);
+    findMdAdditionTasks(tree, file);
+    findRemoveTasks(tree, file);
+  };
+}
+
+function requireFromString(src, filename = 'tmp.js') {
+  const m = new module.constructor();
+  m.paths = module.paths;
+  m._compile(src, filename);
+  return m.exports;
+}
+
+function handleAdditions(tree, action, startIsNode, addNodes) {
+  visit(tree, (node, index, parent) => {
+    if (is(node, startIsNode)) {
+      if (action === 'addMdAfter') {
+        if (node.type === 'root') {
+          node.children.splice(0, 0, ...addNodes);
+        } else {
+          parent.children.splice(index + 1, 0, ...addNodes);
+        }
+      }
+      if (action === 'addMdBefore') {
+        if (node.remarkExtendedProcessed === undefined) {
+          node.remarkExtendedProcessed = true;
```

Comment:
Preventing infinite loops I assume? Because if you add content before, then it will go visit that content, insert before, endlessly like that? If so, needs a comment I think

Reply:
added comment